test/test_persistent_cache.py

```diff
@@ -0,0 +1,187 @@
+import functools
+import os
+from subprocess import run, STDOUT, PIPE
+import sys
+import tempfile
+import unittest
+
+import torch_xla.runtime as xr
+
+
+# Wrapper to manage a temporary directory for the wrapped test
+def run_with_tmpdir(f):
+
+  @functools.wraps(f)
+  def run(*args, **kwargs):
+    with tempfile.TemporaryDirectory() as tmpdir:
+      kwargs.setdefault('tmpdir', tmpdir)
+      f(*args, **kwargs)
+
+  return run
+
+
+# Command to generate a simple graph and perform correctness/metrics assertions.
+# There are four methods supported by configuring the environment for the test:
+#  - Single-device (default): The program runs a computation on a single XLA
+#    device.
+#  - Unsharded SPMD: Setting TEST_WITH_SPMD will run the computation replicated
+#    across all devices.
+#  - Sharded SPMD: Setting TEST_WITH_SPMD and MARK_SHARDING will shard the
+#    computation across all devices.
+#  - Multiprocess: Setting TEST_WITH_MP will run the same computation across all
+#    devices using multiprocessing.
```

Comment:
Why does this test case live in a string? If it has to be a separate file, can it be a file checked into this repository? 

If it's possible, I would actually rather you untangle the different test cases into functions and launch them directly, either with xmp.spawn or normal `multiprocessing`, rather than writing a file to run with subprocess. That may be slightly repetitive, but that's okay for unit tests -- it's better to be very explicit about test assertions than build the best abstraction https://testing.googleblog.com/2019/12/testing-on-toilet-tests-too-dry-make.html

Reply:
I decided to directly use `xmp.spawn` and run SPMD and single-device tests using a ProcessPoolExecutor, so the code lives directly in the test file now. Much cleaner!